### PR TITLE
Prevent underflow/overflow when decoding OGG Vorbis to int16

### PR DESCRIPTION
**Pull request by jantonguirao**
It fixes overflow/underflow issues bug when decoding OGG Vorbis directly to int16.
The applied solution consists on first rounding to int, then clamping the value to the range of int16.

Signed-off-by: Joaquin Anton <janton@nvidia.com>

### CODE REVIEW

File: src/ogg_vorbis.c [hunk 1]
```diff
@@ -552,13 +552,13 @@ vorbis_rshort (SF_PRIVATE *psf, int samples, void *vptr, int off, int channels,
 		float inverse = 1.0 / psf->float_max ;
 		for (j = 0 ; j < samples ; j++)
 			for (n = 0 ; n < channels ; n++)
-				ptr [i++] = psf_lrintf ((pcm [n][j] * inverse) * 32767.0f) ;
+				ptr [i++] = CLAMP_SHORT (psf_lrintf ((pcm [n][j] * inverse) * 32767.0f)) ;
```
Comment (mzient) [inline src/ogg_vorbis.c:555]: Use of a macro will cause double evaluation of the expression - we don't want that.

Comment (mzient) [reply src/ogg_vorbis.c:555]:
```suggestion
                float x = pcm [n][j] * inverse;
				ptr [i++] = psf_lrintf((x < -1 ? -1 : x > 1 ? 1 : x) * 32767.0f);
```

File: src/ogg_vorbis.c [hunk 2]
```diff
@@ -552,13 +552,13 @@ vorbis_rshort (SF_PRIVATE *psf, int samples, void *vptr, int off, int channels,
...
 	}
 	else
 	{
 		for (j = 0 ; j < samples ; j++)
 			for (n = 0 ; n < channels ; n++)
-				ptr [i++] = psf_lrintf (pcm [n][j] * 32767.0f) ;
+				ptr [i++] = CLAMP_SHORT (psf_lrintf (pcm [n][j] * 32767.0f)) ;
```
Comment (mzient) [inline src/ogg_vorbis.c:561]:
```suggestion
		for (n = 0 ; n < channels ; n++) {
				float x = pcm [n][j];
				ptr [i++] = (psf_lrintf ((x < -1 ? -1 : x > 1 ? 1 : x) * 32767.0f)) ;
		}
```

File: src/common.h [hunk 3]
```diff
@@ -113,6 +114,7 @@
 /* Max FLAC sample rate : https://xiph.org/flac/format.html */
 #define		SF_MAX_SAMPLERATE	655350
 
+#define		CLAMP_SHORT(x) (x > SHRT_MAX ? SHRT_MAX : x < SHRT_MIN ? SHRT_MIN : x)
```
Comment (mzient) [inline src/common.h:117]: This isn't necessary when clamping float and it's potentially dangerous and inefficient (it can evaluate x up to 3 times).

**Comment by evpobr** [comment]
Hi @mzient, thanks for pull request. Little bug:

```c
../../src/ogg_vorbis.c:557:21: error: ‘SHRT_MAX’ undeclared (first use in this function)
  557 |     ptr [i++] = x > SHRT_MAX ? SHRT_MAX : x < SHRT_MIN ? SHRT_MIN : x ;
```

**Comment by evpobr** [comment]
And fix code style please.

**Comment by evpobr** [comment]
@arthurt , i've seen similar code for Opus. It turns out that there is a similar problem?

**Comment by jantonguirao** [comment]
> And fix code style please.

I think I fixed the style issues (the errors I got pre-commit). Is there another style-check I need to run?

> @arthurt , i've seen similar code for Opus. It turns out that there is a similar problem?

If Opus can also go outside [-1, 1] range (I am not familiar with that format) then I believe you also have risk of overflow/underflow.

**Comment by arthurt** [comment]
> @arthurt , i've seen similar code for Opus. It turns out that there is a similar problem?

Yes, and all of libsndfile. libsndfile doesn't do clipping by on float->int conversion by default, optimizing for the common case where it doesn't clip. Users can request clipping using SFC_SET_CLIPPING.

Vorbis, Opus and MP3 are treated like a floating point format internally (mostly because that the format the libraries we use do for them).

Vorbis, more than the others, seems to inflate the signal value at points. Vorbis regularly decodes to 1.0, which clips.

See also #194 

Another fix might be to change the default clipping mode when loading Ogg/Opus/MP3 to enable float->int clipping by default.

**Comment by arthurt** [comment]
Side note: @evpobr The SSE float-int conversion routines you added do clipping, always negative. This is what the CPU_CLIPS_NEGATIVE define is about. Allows for one less comparison in the clipping routines (only have to worry about positive clipping).

A useful addition to your changes would be to define CPU_CLIPS_NEGATIVE to 1 if the build configuration decides to use the SSE routines.

**Comment by arthurt** [comment]
Sorry for the comment spam.

This is a know deficiency of the libsndfile api: changing the handling of float->int and int->float conversions are not very symmetric or obvious. You have

- SFC_(GET|SET)_NORM_(FLOAT|DOUBLE)
- SFC_SET_SCALE_FLOAT_INT_READ
- SFC_SET_SCALE_INT_FLOAT_WRITE
- SFC_(GET|SET)_CLIPPING

On top of that, it isn't communicated to users that Vorbis/Opus/MP3 are /like/ float formats, and go through float->int and int->float conversions.

Maybe a more-obvious api and better defaults should be a project goal? Vorbis was the first, and it went unfixed for years. Now Opus and MP3 have come along, it's only going to get worse.

**Comment by jantonguirao** [comment]
@arthurt So, if I understand it correctly, I can configure SFC_SET_CLIPPING and I wouldn't need this PR.
Should I be using SFC_SET_SCALE_FLOAT_INT_READ as well when converting float->int?
Changing the defaults for Vorbis/Opus/MP3 sounds like a good idea.

**Comment by arthurt** [comment]
@jantonguirao So, if I understand it correctly, I can configure SFC_SET_CLIPPING and I wouldn't need this PR.
Should I be using SFC_SET_SCALE_FLOAT_INT_READ as well when converting float->int?

Ideally yes, but I hadn't looked at the ogg_vorbis.c code for a while, and I just looked and saddly no, those commands do no influence it. So that's also an issue.

**Comment by evpobr** [comment]
> Side note: @evpobr The SSE float-int conversion routines you added do clipping, always negative. This is what the CPU_CLIPS_NEGATIVE define is about. Allows for one less comparison in the clipping routines (only have to worry about positive clipping).
> 
> A useful addition to your changes would be to define CPU_CLIPS_NEGATIVE to 1 if the build configuration decides to use the SSE routines (USE_SSE2 is 1).

Are you talking about `psf_lrint` and `psf_lrintf`?

**Comment by arthurt** [comment]
> > Side note: @evpobr The SSE float-int conversion routines you added do clipping, always negative. This is what the CPU_CLIPS_NEGATIVE define is about. Allows for one less comparison in the clipping routines (only have to worry about positive clipping).
> > A useful addition to your changes would be to define CPU_CLIPS_NEGATIVE to 1 if the build configuration decides to use the SSE routines (USE_SSE2 is 1).
> 
> Are you talking about `psf_lrint` and `psf_lrintf`?

Yes, the SSE2 intrinsic instruction `cvtss_si32` clips to the most negative number on all under/over flows.

>The result is rounded according to the current rounding mode. If the result cannot be represented as a 32 bit integer the result will be 0x8000_0000 (i32::MIN) or an invalid operation floating point exception if unmasked (see _mm_setcsr).

This is the behaviour CPU_CLIPS_NEGATIVE is representing :)

**Comment by arthurt** [comment]
I hate to be a butt, but I have an alternate fix, see #747. @jantonguirao thank you for contributing to the project! I don't want this to leave a bad impression, we value contributions!

**Comment by evpobr** [comment]
> Yes, the SSE2 intrinsic instruction `cvtss_si32` clips to the most negative number on all under/over flows.

IMHO it will break clipping to short.

File: src/common.h [hunk 4]
```diff
@@ -996,6 +996,11 @@ static inline int psf_lrint (double x)
 ** Other helper functions.
 */
 
+static inline float clampf (float x, float min, float max)
```
Comment (evpobr) [inline src/common.h:999]: @arthurt , this function is pretty useful.

Comment (arthurt) [reply src/common.h:999]: 1) It operates in the float domain (slow)
2) It's a pretty trivial function
3) It's different from how clipping is handled elsewhere in the project.

**Comment by jantonguirao** [comment]
> I hate to be a butt, but I have an alternate fix, see #747. @jantonguirao thank you for contributing to the project! I don't want this to leave a bad impression, we value contributions!

@arthurt No worries. I am happy that this issue is being handled. I am closing this PR now in favor of #747. Thanks.

**Pull request closed without merging**